[1단계 - 장기] 가콩(최가빈) 미션 제출합니다

gabean13 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

안녕하세요 매트 
이번에 장기 미션 구현한 가콩입니다.
리뷰 잘 부탁드립니다 감사합니다 ☺️

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
[가콩의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4097)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
BoradPiece의 필드가 3개 입니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?
객체 지향적이고, 남이 장기라는 도메인을 잘 몰라도 코드를 이해할 수 있도록 설계하려고 노력하였습니다. 
특히 말(각 말의 규칙을 아는 클래스) -> 장기 말(실제로 장기판 위에서 움직이는 클래스로 좌표와 나라를 알고 있음) <-> 보드(모든 말들을 알고 움직이게 함)의 책임 분리와 관계에 시간을 많이 투자하였습니다. 
리뷰어님이 보시기엔 어떤지 궁금합니다!

이 때, Dynasty(나라)를 Player, Point, BoardPiece 세 개의 클래스에서 알고 있다는 점이 아쉽게 느껴졌습니다. 특히 포인트는 나라를 모르지만 알아야한다는 점(클래스명에서만 명시)에서 휴먼에러?가 발생할 가능성이 크겠다고 느껴졌습니다. 리뷰어님은 이 부분에 대해 어떻게 생각하시는지 궁금합니다. 

규칙을 명시적으로 보여주고 싶어서 경로를 다 쓰거나, 방향을 써 놓았습니다. 경로를 컴퓨터가 찾게하는 코드도 작성을 했었는데요, 그러면 각 말들에 대한 규칙이 코드 내에 있어서 명시적으로 보이지 않는다고 판단해서 해당 방법은 선택하지 않았습니다. 그런데 경로를 직접 입력하니 너무 하드코딩인가 하는 생각과 차나 포 같은 말들은 경로가 없이 방향만 있어서 말들 내부 구현이 달라서 코드를 이해하는데 시간이 걸리지 않을까 하는 생각이 들었습니다. 혹시 리뷰어님은 두가지 방법 중에 어떤 것을 선택하셨을 지 궁금합니다.
규칙을 작성한 코드입니다. (test코드로 작성하여 많이 허술한 부분 죄송합니다.)
```java
    @Test
    void 나이트이동경로() {
        // knight
        //경로를 컴퓨터가 찾기
        Point start = new Point(3, 2);
        Point end = new Point(1, 3);

        // 단위 벡터를 계산한다
        Point interval = end.minus(start);
        int unitX = interval.x() / Math.abs(interval.x());
        int unitY = interval.y() / Math.abs(interval.y());
        Point unitVector = new Point(unitX, unitY);

        //먼저 대각선으로 이동
        List<Point> path = new ArrayList<>();
        Point prePoint = end.minus(unitVector);
        path.add(prePoint);

        for (Point point : path) {
            System.out.println("point = " + point.x() + ", " + point.y());
        }
    }

    @Test
    void 상이동경로() {
        // elephant
        //경로를 컴퓨터가 찾기
        Point start = new Point(3, 3);
        Point end = new Point(6, 5);

        // 단위 벡터를 계산한다
        Point interval = end.minus(start);
        int unitX = interval.x() / Math.abs(interval.x());
        int unitY = interval.y() / Math.abs(interval.y());
        Point unitVector = new Point(unitX, unitY);

        //대각선으로 이동
        List<Point> path = new ArrayList<>();
        Point prePoint1 = end.minus(unitVector);
        path.add(prePoint1);
        Point prePoint2 = prePoint1.minus(unitVector);
        path.add(prePoint2);

        for (Point point : path) {
            System.out.println("point = " + point.x() + ", " + point.y());
        }
    }
```

마지막으로 개인적인 궁금증인데요, TDD 좋아하시고 애용하시는지 궁금합니다. 저는 아직 TDD가 불편하고, 감을 못찾은 것 같습니다 🥲 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **문서**
  - README가 업데이트되어 게임의 규칙, 진행 방식, 플레이어 등록 및 보드 설정 가이드가 추가되었습니다. 한국어 제목과 상세한 요구사항으로 게임 이해도를 높입니다.

- **새로운 기능**
  - 인터랙티브한 게임 엔진이 도입되었습니다. 사용자 입력 검증, 보드 시각화 및 순차적 게임 진행이 개선되어 플레이어가 직관적인 게임 환경을 경험할 수 있습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gabean13 committed:
feat: 폰 이동 기능 구현

gabean13 committed:
feat: 장기판 상차림에 따라 초기화 기능 구현

gabean13 committed:
feat: Knight가 목적지로 갈 수 있는지 판단하는 기능 구현

gabean13 committed:
feat: 한나라, 초나라 별로 Point 구현 및, 방향 구현

gabean13 committed:
feat: Board위에서 위치를 가지는 PointPiece 구현

gabean13 committed:
test: PointPiece, Pawn, JanggiBoard 단위 테스트 작성

gabean13 committed:
feat: Knight 이동 구현

gabean13 committed:
feat: Elephant 이동 구현

gabean13 committed:
feat: Rook 이동 구현

gabean13 committed:
feat: Cannon 이동 구현

gabean13 committed:
feat: 게임 초기화 구현

- 플레이어 이름 입력 및 저장
- 상차림 입력 및 보드판 초기화
- 초기 보드판 출력

gabean13 committed:
feat: 플레이어가 번갈아가며 보드판 움직임 구현

gabean13 committed:
refactor: 말 싱글톤 생성자 삭제 및 equals, hashcode 구현

gabean13 committed:
refactor: Point를 추상 클래스로 변경

gabean13 committed:
refactor: 메서드 의미있는 단위로 분리

gabean13 committed:
rename: 장기 말 클래스 명 변경

gabean13 committed:
feat: 플레이 순서 초나라 부터 시작하게 변경 및 UI 개선

gabean13 committed:
feat: 입력 형식이 틀린경우 재입력 기능 구현

gabean13 committed:
test: Point 테스트 작성 및 테스트 명 추가

gabean13 committed:
docs: 기능 요구사항 수정

hyeonic commented on src/main/java/janggi/domain/piece/Chariot.java:
```diff
@@ -0,0 +1,55 @@
+            Direction.UP, Direction.DOWN, Direction.RIGHT, Direction.LEFT
+    );
+
+    private static boolean canMoveEndPointByDirection(JanggiBoard janggiBoard, Point start, Point end,
+                                                      Direction direction) {
+        Point currPoint = start;
+        while (canMoveUntilEndPoint(end, currPoint)) {
+            currPoint = currPoint.move(direction);
+            if (janggiBoard.isExistPiece(currPoint)) {
+                break;
+            }
+        }
+        return currPoint.isSamePosition(end);
+    }
+
+    private static boolean canMoveUntilEndPoint(Point end, Point currPoint) {
+        return !currPoint.isSamePosition(end) && currPoint.isNotOutOfBoundary();
+    }
```
static 메서드로 활용한 이유가 있을까요?

gabean13 replied:
extract method 하다가 실수했습니다ㅠㅠ 죄송합니다. 더 꼼꼼히 확인하도록 하겠습니다. 🥲

hyeonic commented on src/test/java/janggi/domain/piece/ChariotTest.java:
```diff
@@ -0,0 +1,86 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+class ChariotTest {
```
꼼꼼하게 테스트를 작성하셨네요 👍

hyeonic commented on README.md:
```diff
@@ -1,3 +1,116 @@
+> 장기 미션 저장소
 
-장기 미션 저장소
+### 요구사항
```
요구사항을 잘 정의하였네요 😃

추가로 말 규칙에 체스 관련 언어가 아닌 실제로 가콩이 설계한 장기의 객체 이름도 같이 매칭 시키면 이해하는데 더욱 용이할 것 같아요!

gabean13 replied:
넵 클래스명을 중간에 고치느라 이부분을 놓쳤네요 ㅠㅠ 수정했습니다!!

hyeonic commented on src/test/java/janggi/domain/piece/ChariotTest.java:
```diff
@@ -0,0 +1,86 @@
+
+        // then
+        assertThat(isMovable).isTrue();
+    }
```
설계한 이름에 맞춰 변수명도 `chariot`로 통일할 순 없을까요?

gabean13 replied:
넵 중간에 클래스명을 바꾸면서 놓친 것 같습니다. 죄송합니다 수정했습니다!!

hyeonic commented on src/main/java/janggi/domain/piece/Chariot.java:
```diff
@@ -0,0 +1,55 @@
+import janggi.domain.board.point.Point;
+import java.util.Set;
+
+public class Chariot implements Piece {
```
가콩은 어떠한 기준으로 메서드 순서를 배치하나요?

gabean13 replied:
사실 아직 딱 저만의 메서드 배치 기준은 없는 것 같습니다..ㅜㅜ 페어들의 다양한 방식을 만나면서 시도해보고 있습니다!
기존에는 생성자 / public / 그 아래 사용되는 private / public ... / object override메서드 / getter 이런 순서로 배치를 했었는데, 이번에는 생성자 / public / private  / object override메서드 / getter 순서로 배치를 하려고 했습니다!

원래 방식이 코드를 읽을 때 흐름이 자연스러워서 읽기 편할 것 이라고 생각을 했었기에 그런식으로 배치를 했었는데요, 

만약 코드가 꽤 길어졌을 때, public 메서드가 너무 아래로 밀려서 오히려 더 메서드를 찾기 힘들어질수도 있겟다는 점과 
여러 public 메서드에서 사용되는 private은 배치하기 애매하다는 점 등을 고려하여 
이번에는 public/private 방식으로 배치를 했습니다.

매트는 어떤 방식을 선호하는지 여쭤봐도될까요?

hyeonic replied:
저는 보통 public과 해당 public 메서드에서 사용하는 private을 두는 방향을 선호해요. 자연스럽게 코드 흐름을 따라가는 장점이 있더라구요! 

어떤 방법이든 애플리케이션 내에 일관성만 존재한다면 상관 없다고 생각합니다 ㅎㅎ

gabean13 replied:
넵, 저도 사실 그 배치를 더 선호합니다ㅎㅎ😊 
그런데 이 방식에서 가장 어렵다고 느낀 부분이 여러 public 메서드에 사용되는 private메서드의 위치였습니다. 
혹시 매트는 어디에 배치되었을 때 읽기 편하셨나요?!?!

hyeonic replied:
저는 가장 하단의 public 아래에 두는 것이 읽기 좋더라구요 ㅎㅎ

hyeonic commented on src/main/java/janggi/domain/piece/Elephant.java:
```diff
@@ -0,0 +1,61 @@
+            }
+        }
+        return false;
+    }
```
가능하다면 뎁스를 좀 더 줄여볼까요? stream에서 제공하는 다양한 메서드(anyMatch, reduce 등)를 고려해보면 좋을 것 같아요!

gabean13 replied:
넵, stream의 anyMatch를 사용해서 indent를 줄였습니다! 
그런데 다른 indent가 2인 메서드들은 중간에 break가 있어서 indent를 줄이지 못했습니다. 스트림에서 다양한 메서드를 조합하면 해결할 수 있긴하나, 가독성이 너무 저하된다고 느껴서 indent가 2이더라도 그대로 두는 게 낫다고 생각했습니다. 
이것에 대한 매트의 생각을 여쭈어봐도 될까요? 🥺

hyeonic replied:
넵 가능한 케이스에 최대한 활용하는 편이 좋다고 생각해요 👍

hyeonic commented on src/main/java/janggi/view/InitializeView.java:
```diff
@@ -0,0 +1,63 @@
+            return ChuBoardSetUp.OUTER_ELEPHANT;
+        }
+        return ChuBoardSetUp.INNER_ELEPHANT;
+    }
```
반복된 if문을 줄여볼 순 없을까요?

gabean13 replied:
넵, Setup 을 하나로 통일하고, switch문으로 변경했습니다. 
개인적인 생각으로 이 경우에 case에 해당하지 않으면 default로 예외를 던지는 부분이 명확하다고 생각하여서 switch를 사용했습니다.

hyeonic replied:
넵 좋습니다! 다만 상차림에 대한 방식이 추가되거나 제거되면 어떻게 될까요? 왜 우리는 반복된 if나 switch 사용을 지양해야 할까요?

gabean13 replied:
넵, 
반복된 if 나 switch문을 쓰는 경우라면 `어떤 값이면 무엇을 선택해야한다`는 행동을 하는 메서드일 가능성이 높습니다. 
결국 새로운 값이 추가될 때 마다 if나 swtich를 추가해야줘야하고 이는 변경에 열려있는 경우라고 볼 수 있습니다. 
이는 유지보수가 어려울 것 같습니다.

그래서 BoardSetUpMenu라는 class를 만들었습니다.  
BoardSetUp의 field에 도메인 관련 값들을 추가할까 하는 생각도 했었는데요, 그렇게 된다면 뷰가 변경될 때마다 도메인을 변경시켜야한다는 큰 단점때문에 새로운 클래스를 만들었습니다. 
이 클래스에서는 BoardSetUp의 뷰 관련 값들을 가지고 입력에 맞는 BoardSetUp을 반환합니다.

이렇게 하면 값이 추가되거나 변경되었을 때, 관련없는 View를 수정시키는게아니라 BoardSetUpMenu와 BoardSetUp만 변경 하면 됩니다. 
이때 하나의 값을 변경하더라도, 두개의 클래스를 유지보수해야한다는 단점이 있갰다고 생각을 했습니다. 
하지만 도메인이 뷰를 아는게 더 큰 단점이라고 판단하여 이렇게 리팩토링하였습니다.

이 방법에 대해 매트의 생각은 어떠신지 여쭤봐도될까요??🥺
코멘트와 관련된 커밋 : [01d343a](https://github.com/woowacourse/java-janggi/pull/18/commits/01d343aec12d7002c9331b04a314ae68d584fb54)

hyeonic replied:
> 도메인이 뷰를 아는게 더 큰 단점이라고 판단하여 이렇게 리팩토링하였습니다.

저도 가콩과 비슷한 관점을 가지고 있어요 ㅎㅎ 잘 수정하신 것 같습니다 😃

hyeonic commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.Optional;
+import java.util.Set;
+
+public class JanggiBoard {
```
> 객체 지향적이고, 남이 장기라는 도메인을 잘 몰라도 코드를 이해할 수 있도록 설계하려고 노력하였습니다.
특히 말(각 말의 규칙을 아는 클래스) -> 장기 말(실제로 장기판 위에서 움직이는 클래스로 좌표와 나라를 알고 있음) <-> 보드(모든 말들을 알고 움직이게 함)의 책임 분리와 관계에 시간을 많이 투자하였습니다.
리뷰어님이 보시기엔 어떤지 궁금합니다!

보았을 때 적절히 책임이 분리된 것 같아서 코드를 이해하는데 큰 어려움은 없었습니다 👍

hyeonic commented on src/main/java/janggi/domain/board/point/ChuPoint.java:
```diff
@@ -0,0 +1,24 @@
+
+import janggi.domain.board.Direction;
+
+public class ChuPoint extends Point {
```
> 이 때, Dynasty(나라)를 Player, Point, BoardPiece 세 개의 클래스에서 알고 있다는 점이 아쉽게 느껴졌습니다. 특히 포인트는 나라를 모르지만 알아야한다는 점(클래스명에서만 명시)에서 휴먼에러?가 발생할 가능성이 크겠다고 느껴졌습니다. 리뷰어님은 이 부분에 대해 어떻게 생각하시는지 궁금합니다.

답변하기에 앞서 가콩은 어떤 목적으로 각 진영의 포인트 로직을 구분하신 건가요??

gabean13 replied:
일단 이런 로직을 구현하게된 시작은 졸(Solider)이었습니다. 졸은 자기 나라 기준으로 앞 , 좌, 우로만 이동할 수 있어서 이를 한 Solider로 어떻게 구현할까 하다가 나라별로 바라보는 방향에 따라 up, down, right, left를 정의하면 한 solider 클래스로 구현할 수 있겟다고 생각하여 이런 식으로 구분하게 되었습니다.

이후에 코드를 리팩토링하면서 point부분이 과하게 추상화가 되었다고 느꼈습니다. point와 dynasty가 굉장히 강하게 결합되어있는데 막상 코드에서는 드러나는 부분이 없어서 둘이 불일치가 되어도 모르는 경우가 많았습니다. 
또, 좌표상에서 up down이 나라에 따라 다르다는게 다시 생각해보니 굉장히 어색하다고 느꼈습니다.

따라서 Point와 Setup을 모두 하나로 통일시키고 문제가 되었던 Solider만 Han, Chu에 따라 다르게 좌표를 가지도록 변경하였습니다.
코드가 훨씬 간결해지고(코드 양 자체가 감소 + 나라별로 up,down,left,right ,, 에 대해서 알지 않아도됨), 나라와 포인트가 불일치 될 일이 없고, 좌표상 움직임이 항상 동일하니 더 코드를 이해하기 쉽다고 느꼈습니다

이때 아쉬운 점은 Solider의 인스턴스는 결국 Han 아니면 Chu 한가지만 사용할텐데 더 클래스를 분리해야하나? 하는 생각이 들기도 했습니다. 매트의 생각은 어떠신지 궁금합니다!😃

hyeonic replied:
Solider에서 `공유하는 부분`과 `변경되는 부분`은 어디일까요? dynasty와 그에 따른 이동 전략이 다르다면 해당 정보를 외부에서 주입 받을 수도 있지 않을까요?

gabean13 replied:
다른 부분은 Path를 제외하고는 다 동일합니다!
그래서 Solider라는 추상 클래스를 만들고 HanSolider와 ChuSolider라는 구현체를 만들었습니다. 
관련 커밋 : https://github.com/woowacourse/java-janggi/pull/18/commits/503d8f1c327544d14ee1ac702bb904479cf5cf8b

hyeonic commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static final Set<Direction> DIRECTIONS = Set.of(
+            Direction.UP, Direction.DOWN, Direction.RIGHT, Direction.LEFT
+    );
```
> 규칙을 명시적으로 보여주고 싶어서 경로를 다 쓰거나, 방향을 써 놓았습니다. 경로를 컴퓨터가 찾게하는 코드도 작성을 했었는데요, 그러면 각 말들에 대한 규칙이 코드 내에 있어서 명시적으로 보이지 않는다고 판단해서 해당 방법은 선택하지 않았습니다. 그런데 경로를 직접 입력하니 너무 하드코딩인가 하는 생각과 차나 포 같은 말들은 경로가 없이 방향만 있어서 말들 내부 구현이 달라서 코드를 이해하는데 시간이 걸리지 않을까 하는 생각이 들었습니다. 혹시 리뷰어님은 두가지 방법 중에 어떤 것을 선택하셨을 지 궁금합니다.
규칙을 작성한 코드입니다. 

아마 이런 부분을 말씀하신 거겠죠? 아니라면 다시 질문주세요. 저도 가콩이 하신 방법이 각 말들이 움직일 수 있는 방향에 대해 명확하게 알고 있기 때문에 더 좋은 방향이라고 생각했어요. 다만 방향만 알고 있다보니 결국 각 말이 갈 수 있는 방법은 `isMovable`에 모이게 되어 명확하게 어떠한 `이동 전략`을 가질 수 있는지 확인하기 어려운 것 같아요. 이 부분은 어떻게 개선할 수 있을까요?

gabean13 replied:
넵, 각 말들이 갈 수 있는 방법을 필드로 알도록 이동전략 인터페이스를 만들어서 구현했습니다.
- LongRange : 방향만 알고, 명확한 경로는 존재하지 않는 말들 (Chariot)
- FixedRange : 경로가 고정되어있는 말들(Horse, Elephant, Solider)
- Cannon : 포의 움직임

이렇게 하면 말들이 각 방향만 알면, 이동 전략을 통해 움직일 수 있습니다.
- 이 방식이면 3가지 이동전략만 알면 이동 방법을 이해할 수 있겠다는 장점이 있습니다.
- 클래스 명으로 이동 전략을 명시하니 좀 더 이해가 쉬울 것 같습니다.  
- 다른 전략이 생긴다면 쉽게 추가할 수 있습니다.
- 중복되는 이동전략 코드를 없앨 수 있습니다.

단점으로는
- 코드의 복잡도가 더 올라갑니다.

혹시 매트는 이 코드로 가독성이 개선됨을 느끼셨는지 아니면 오히려 복잡하다고 느끼셨는지 궁금합니다!

hyeonic replied:
이동 전략에 대한 추상화를 통해 각각의 구체적인 전략이 구현체로 잘 이동된 것 같아요 👍

한 단계 더 나아가 각각의 기물은 N개의 이동 전략을 가질 수도 있을 것 같아요. 포를 통해 예를 들어볼게요.
1. 포는 목적지 사이에 포를 제외한 기물이 있을 때 이동 가능하다.
2. 포는 목적지에 있는 기물이 적군 기물이어야 이동 가능하다.

이렇게 세분화 하게 된다면 좀 더 작은 단위에 집중하여 기물의 이동 전략을 다룰 수도 있을 것 같습니다. 요건 참고만 해주세요~

gabean13 replied:
오 넵!! 
그런식으로는 생각해보지 못했는데, 만약 복잡한 기물들이 계속 생긴다면 코드를 확장하고 재사용 할 때 편리할 것같습니다. 
또 지금보다 이동 전략을 이해하기도 더 쉬울 것 같아요!!
좋은 방향 알려주서셔 감사합니다 ☺️

hyeonic reviewed (changes requested):
안녕하세요 가콩~ 장기 리뷰를 맡은 매트입니다 😃 쉽지 않은 요구사항이었을 땐데 꼼꼼히 정리해주셔서 룰을 이해하는데 큰 어려움은 없었습니다. 간단히 코멘트 남겼으니 확인해주시고, 질문이 어렵거나 모호한 부분 있다면 언제든 추가 코멘트 혹은 DM 주세요!

---

> 마지막으로 개인적인 궁금증인데요, TDD 좋아하시고 애용하시는지 궁금합니다. 저는 아직 TDD가 불편하고, 감을 못찾은 것 같습니다 🥲

아직 익숙하지 않아 더욱 큰 거부감이 들 수 있다고 생각해요. 모든 경우에 TDD를 활용하는 것은 아니지만 보통 여러 객체가 상호작용해야 할 때 객체에 적절한 책임을 찾기 위해 TDD를 활용하곤 합니다. TDD를 통해 해소하고자 하는 메시지를 의식적으로 던지다보면 그것을 소화할 수 있는 객체가 보이더라구요 ㅎㅎ 먼저 가콩 잘 정리하신 것 처럼 작은 단위로 요구사항을 잘 구분하는 것 부터 시작이라고 생각해요. 하나씩 체크 리스트를 채워가면서 의식적으로 연습하다보면 자연스럽게 객체의 책임을 찾을 수 있지 않을까 합니다!

gabean13 committed:
refactor: 불필요한 static 삭제

gabean13 committed:
refactor: indent 줄임

gabean13 committed:
refactor: BoardSetup 한 클래스로 관리하도록 변경

gabean13 committed:
refactor: point 삭제, position 기능 구현

- 나라별로 point 가지는 것 삭제
- position으로 통일

gabean13 committed:
refactor: 보드 자료구조 Set에서 Map으로 변경 및 보드피스는 좌표를 모르도록 변경

gabean13 committed:
"test: 상차림 테스트 추가"

gabean13 committed:
feat: Direction 기능 구현

- Direction에 따라 움직인 Position을 반환
- 해당 Direction으로 움직일 수 있는지 판단
- Point는 좌표(보드내)의 역할만 수행하도록 변경

gabean13 committed:
feat: 목적지에 같은 나라가 있는지 검사 기능 구현 및 test 추가

- 메서드 순서 변경
- Optional 삭제

gabean13 committed:
feat: 이동경로 기능 구현

- 포 / 고정경로 / 경로고정X 로 나누어서 구현

gabean13 committed:
refactor: BoardPiece가 목적지에 같은 나라가 있는지 확인하는 책임 제거 및 테스트 추가

gabean13 committed:
refactor: 각 기물에 따른 이동전략 추가 및 테스트 작성

gabean13 committed:
rename: player 패키지 생성 및 클래스 이동

gabean13 committed:
refactor: 코드 정렬 및 메서드 순서 변경

gabean13 commented:
안녕하세요 매트😆
코멘트에도 작성했지만 크게 변한 부분들이 많아서 여기에도 변한 부분에 대한 설명 추가했습니다.!

ChuPoint, HanPoint, Default Point / ChuBoardSetUp, HanBoardSetUp 삭제
- point와 setup은 dynasty와 굉장히 강한 결합인데 이게 명시적으로 주입해준게 아니라, 클래스명으로만 알고 컴퓨터에서는 알지 못하므로 예외가 발생했을 때 찾기가 매우 힘들었습니다. 둘이 불일치 되었을 때 이를 예외처리하지 못하는 부분도 단점이라고 생각했습니다. 단점이 장점보다 훨씬 크다고 판단하여 하나로 통일하였습니다. 
- 이에, Point, BoardSetup으로 통일하였습니다.

JanggiBoard가 Set<BoardPiece> -> Map<Point, BoardPiece>으로 변경 및 BoardPiece에서 Position필드 삭제
- 한 Postion에는 반드시 한개의 BoardPicece만 가져야하므로 Map이 더 적절한 자료구조라고 생각했습니다.
- BoardPiece가 Point을 가지면서 불변객체로 만들 수가 없었습니다.
- 따라서 위치를 키값으로 가지는 Map을 janggiboard가 관리하고, boardpiece는 나라만 알도록 변경하였습니다.

움직임 전략을 추가하였습니다.
- 각 말들이 움직임전략을 가지고 있습니다.
- LongRange : 방향만 알고, 명확한 경로는 존재하지 않는 말들 (Chariot)
- FixedRange : 경로가 고정되어있는 말들(Horse, Elephant, Solider)
- Cannon : 포의 움직임

추가적으로 테스트들을 추가하였습니다.

**아쉬운 점**
인터페이스 시그니처를 통일시켜야해서 
- MovementStrategy에서 방향만 가지는 long, cannon에서 List<Direction> 으로 Direction을 표현했습니다.
- Piece에서 Soldier를 위해 Dynasty를 다른 객체에서도 다 받아야합니다.(다른객체에는 사용하지않음)
억지로 끼워맞추었습니다

이 부분은 아직 방법을 찾지 못해서 추후에 개선하도록 하겠습니다ㅠㅠ🥲

hyeonic commented on src/main/java/janggi/domain/piece/Cannon.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+import java.util.Set;
+
+public class Cannon implements Piece {
+
+    private static final Set<List<Direction>> DIRECTIONS = Set.of(
+            List.of(Direction.UP), List.of(Direction.DOWN), List.of(Direction.RIGHT), List.of(Direction.LEFT)
+    );
+
+    private final MoveStrategy moveStrategy = new CannonMoveStrategy();
+
+    @Override
+    public boolean isMovable(JanggiBoard janggiBoard, Dynasty dynasty, Point start, Point end) {
+        return moveStrategy.isMovable(janggiBoard, start, end, DIRECTIONS);
+    }
```
MoveStrategy를 통해 이동 전략을 인터페이스로 구성하였네요! 구현체를 직접 생성하여 관리한다면 필드로 두지 않고 상수로 관리해도 될 것 같아요. Connon 입장에서 MoveStrategy를 인터페이스를 필드로 의존할 때 어떠한 이점을 기대할 수 있을까요? 

다른 관점으로는 Piece를 구현하는 대부분의 객체들이 MoveStrategy을 가질 것으로 보여요. 상속 받는 객체에게 이러한 필드를 가질 수 있도록 강제할 순 없을까요?

gabean13 replied:
해당 moveStraetgy의 구현체를 직접 생성하여 필드로 둔 이유는 장기가 규칙이 바뀌지 않을 게임이라고 생각을 해서 생성자 주입 대신 바로 할당을 했습니다. 
직접 생성하여 관리한다면 상수로 관리하는 것이 낫겟네요!! 

movestrategy를 필드로 의존하고 의존성을 주입해준다면, 장기에서는 큰 이점이 없겠으나 움직임을 클라이언트가 원하는데로 갈아낄 수 있다는 장점이 있을 것 같습니다.

상속받는 객체인 Piece들의 구현체가 동일하게 가지는 MoveStrategy와 Set<List<Direction>> 필드와 isMovable 메서드를 가지도록 Piece 인터페이스를 구현하는 추상 클래스 PieceAbstractInterface를 만들었습니다.
이 클래스로 Piece의 구조를 강제로 만들어 줄 수 있습니다.

hyeonic replied:
좋습니다! 다만 PieceAbstractInterface없이 Piece를 추상 클래스로 두어도 괜찮을 것 같아요. 현재는 PieceAbstractInterface를 추상 클래스로 두는 것이 어떠한 이점이 있는지 명확하게 이해하기가 어려운 것 같아요.

hyeonic commented on src/main/java/janggi/domain/piece/BoardPiece.java:
```diff
@@ -0,0 +1,58 @@
+import janggi.domain.board.Point;
+import java.util.Objects;
+
+public class BoardPiece {
+
+    private final Piece piece;
+    private final Dynasty dynasty;
+
+    public BoardPiece(Piece piece, Dynasty dynasty) {
+        this.piece = piece;
+        this.dynasty = dynasty;
+    }
```
dynasty를 밖에서 구성하신 이유가 있을까요? Piece가 해당 정보를 가져도 괜찮지 않을까요?

gabean13 replied:
제가 생각했을 때,
Piece는 규칙에 따른 움직임만 알고 움직일 수 있는지만 판단하는 책임이 있고,
Board위로 올라가서 BoardPiece가 되어서 Piece와 Dynasty를 알고 나라 관련 판단을 하는 책임이 있는게 더 현실과 비슷하다고 생각을 했습니다. 

혹시 매트는 어떻게 생각하셨는지 궁금합니다.🥹

hyeonic replied:
가콩이 의도한 방향이 그렇다면 괜찮다고 생각해요 👍

hyeonic commented on src/main/java/janggi/domain/board/Point.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Objects;
+
+public record Point(
+        int x, int y
+) {
```
가콩은 어떠한 상황에서 record를 고려하시나요? 일반적인 class와 어떠한 차이가 있을까요?

gabean13 replied:
저는 record를 불변 데이터값을 포장하고 싶을 때 주로 사용합니다.
Point에 record를 사용한 이유는 Point에게 좌표!라는 역할만 부여하고 따로 추가적인 비지니스 로직을 주고 싶지 않았기 때문입니다.

record와 일반적인 calss의 차이점은
- record는 불변 클래스임을 보장해줍니다. 
- record는 보일러플레이트 코드(생성자, equals, toString, getter)를 개발자가 쓰지 않아도 컴파일 시점에 생성해줍니다. 
대신 원치 않게 필드를 노출시킬수도 있으니 간단한 불변 객체를 만들고 싶을 때만 사용하는 편입니다.

hyeonic replied:
좋습니다! 불변 객체를 활용 했을 때 어떤 이점이 있을까요?

hyeonic reviewed (changes requested):
안녕하세요 가콩! 질문 주신 부분 코멘트 남겼습니다. 확인해주시고 이해가 어려운 부분이 있다면 언제든 편하게 DM 남겨주세요~

hyeonic commented on src/main/java/janggi/domain/piece/moveStrategy/CannonMoveStrategy.java:
```diff
@@ -0,0 +1,47 @@
+        return isJump && current.equals(end) && !janggiBoard.isExistCannon(end);
+    }
+
+    private static boolean isNotEnd(Point current, Point end) {
+        return !current.equals(end);
+    }
```
요 부분도 static은 불필요해보이네요!

gabean13 replied:
헉 넵 감사합니다 수정했습니다ㅠㅠ

gabean13 committed:
refactor: Soldier 추상 클래스 & Chu, Han이 이를 구현하도록 변경

gabean13 committed:
refactor: Piece interface 시그니처 변경

gabean13 committed:
refactor: BoardSetUp에 대한 View를 관리하는 BoardSetUpMenu 추가

gabean13 commented:
@coderabbitai full review

gabean13 committed:
refactor: Piece를 구현하는 추상골격클래스 구현 및 말들이 추상 골격클래스를 구현하도록 변경

gabean13 committed:
refactor: BoardPiece가 자신의 나라 기물이 아닌지 판단하도록 변경

hyeonic commented on src/main/java/janggi/domain/piece/Piece.java:
```diff
@@ -0,0 +1,9 @@
+import janggi.domain.board.JanggiBoard;
+import janggi.domain.board.Point;
+
+public interface Piece {
+
+    boolean isMovable(JanggiBoard janggiBoard, Point start, Point end);
+}
```
Piece와 JanggiBoard가 상호 참조하는 구조가 될 수 있을 것 같아요. 상호 참조는 어떤 문제를 야기할 수 있는지, 어떻게 하면 이러한 상호 참조를 제거할 수 있을까요?

hyeonic commented on src/main/java/janggi/domain/piece/General.java:
```diff
@@ -0,0 +1,8 @@
+package janggi.domain.piece;
+
+public class General extends PieceAbstractInterface {
+
+    public General() {
+        super(null, null);
+    }
+}
```
이후 단계에서 고려될 것 같은데 null보다 `Collections.emptySet()`이나 아무 이동도 하지 않는 구현체를 두는 방향은 어떨까요??

hyeonic reviewed (approved):
안녕하세요 가콩! 이전보다 훨씬 개선된 것 같아서 머지해도 될 것 같아요 ㅎㅎ 간단히 코멘트 남겼으니 이후 단계에서 같이 고려해주세요!
